### ISSUE TITLE
Create stream/table index functions

### ISSUE THREAD

**Issue by nathanielc** [influxdata/flux#1089]
We need a class of functions that allows a user to index into a stream of tables and the tables themselves to select either a scalar value, a row record or a column slice.

These functions will depend on #1088.

We likely need at least these three functions:

```
selectTable // selects a specific table from a stream
sliceColumn // given a single table produce an array of column values
sliceRow // given a single table produce a record for the row
scalarValue // given a table produce the scalar value at a specific cell in the table
```

We need to design out these functions, using numeric indexes is likely prone to creating brittle code as it will depend on table, column and row order being stable. We should use map lookup like interfaces where possible, for example the sliceColumn function should accept a column label name, not a numerical index. For tables inside a stream we should use the group key concept. For rows I can't think of anything but a numeric index or a predicate search function.

The above functions names are not final, when we have a design that works with the expected types we can determine better names.

The way I expect these function to work is once #1088 is complete then these function can consume a table object as an argument, execute the table object and use that as the input value. Maybe on the selectTable needs to do that part while the rest of the functions operate on pure values.Value types. Should we create easy to use implementation of a `table` type that represents a table as a list of records?

Also note that we will eventually want syntax sugar for these function but for now that work is out of scope of this issue.

This is part of #1085

**Comment by nathanielc** [comment]
I think we only need three functions for now:

```
// tableFind returns the first table that matches the predicate function
builtin tableFind : (tables : <-stream, fn: (key) -> bool ) -> table

// columSlice returns a list of values in a column of a table
builtin columnSlice : (table : <- table, column : string) -> []'a

// record returns a specific row from a table as a record
builtin record : (table : <- table, row: int) -> record
```

To get a scalar value from a specific cell use `record` and then access the desired property directly:

```
from() |> ... |> tableFind(fn: (key) => true) |> record(row: 0)._value
```

The `topN` use case looks like this:

```
hosts = from() 
   |> ... 
   // group everything into a single table
   |> group() 
   // find the first table (and only table)
   |> tableFind(fn: (key) => true) 
   // get list of hosts from table
   |> columnSlice(column:"host")

from()
   // use list of hosts in filter condition
   |> filter(fn:(r) => r.host in hosts)
```

**Comment by nathanielc** [comment]
As for implementing these functions I see the break down being as follows:

1. The `tableFind` function is responsible for compiling and executing the TableObject.
2. The `tableFind` object then returns the single table as a values.Value 
3. Both `columnSlice` and `record` consume the values.Value and type assert it to a flux.Table and then perform the rest of their function.

I do not want to create a new semantic type for tables. I am fine of the implementation of values.Value for a table implements a fake values.Object with no properties. The reason I do not want to create a new semantic type for tables is that the tables type is really a composite type of other base types and so it does not need its own type. Creating the special table type just means we have to remove it later. 

These are just my thoughts on implementation. If you see a different way please feel free to suggest it.

**Comment by jlapacik** [comment]
I'm putting this back on the backlog for right now. I think this work should wait until the `refactor/private-spec` branch has been merged to master.

**Comment by affo** [comment]
I assigned this to myself because I thought I could pick it today and I would really like to pick it 😅 .
However I still need to finish up some task, so I cleared assignees.

**Comment by jsternberg** [comment]
> I do not want to create a new semantic type for tables. I am fine of the implementation of values.Value for a table implements a fake values.Object with no properties. The reason I do not want to create a new semantic type for tables is that the tables type is really a composite type of other base types and so it does not need its own type. Creating the special table type just means we have to remove it later.

I'm not sure how I feel about this. I think if we're going to do this, then it needs to be in the spec and we need to define the schema for what kind of object a table is. Further, this may be problematic since I don't think a table can necessarily be represented completely by an object.

If we decide to do things this way, then the table operations will have to operate on an object with a specific schema. That schema will have to expose all of the properties of the table with both the table schema and the values within the table so that we can retrieve those values.

Are we ok with doing that? If we have a specific table type, then it'll be easier for the type checker to say "yes, I have a table" and we won't be able to confuse a table with an object since an object is also used for a specific record.

**Comment by nathanielc** [comment]
> If we decide to do things this way, then the table operations will have to operate on an object with a specific schema

I don't follow this logic, what is the reason that we will be tied to a specific schema?

**Comment by jsternberg** [comment]
If a table is an object, it will have a type definition because it will be an object. That's what I mean by a schema. If we use #463 as an example, the schema for the table type would be an object with a `schema` property and a `records` property. The values for those properties will also have their own types that are part of the type definition.

I hadn't seen #463 previously so I've got no opposition to doing it that way, but we should probably specify what that would look like sooner rather than informally selecting the values currently and having to change them later.

**Comment by affo** [comment]
At the moment, I am implementing a Table type that only exposes the schema.

Records are available by using the function `record` and columns by the function `columnSlice` (though I'd prefer a better name for it).

Having a Table type could be a good thing in that we can be compile-time safe when checking that `records(table: table, i: 0)` is really provided with a table.

I'll provide a first implementation in the associated PR #1174 so that you can comment on it.
